x-pack/legacy/plugins/uptime/server/lib/adapters/monitor_states/elasticsearch_monitor_states_adapter.ts

```diff
@@ -29,11 +29,119 @@ interface LegacyMonitorStatesQueryResult {
   afterKey: any | null;
 }
 
+interface LegacyMonitorStatesRecentCheckGroupsQueryResult {
+  checkGroups: string[];
+  afterKey: any | null;
+}
+
 export class ElasticsearchMonitorStatesAdapter implements UMMonitorStatesAdapter {
   constructor(private readonly database: DatabaseAdapter) {
     this.database = database;
   }
 
+  // This query returns the most recent check groups for a given
+  // monitor ID.
+  private async runLegacyMonitorStatesRecentCheckGroupsQuery(
+    request: any,
+    query: any,
+    searchAfter?: any
+  ): Promise<LegacyMonitorStatesRecentCheckGroupsQueryResult> {
+    const body = {
+      query: {
+        bool: {
+          filter: [
+            {
+              // We check for summary.up to ensure that the check group
+              // is complete. Summary fields are only present on
+              // completed check groups.
+              exists: {
+                field: 'summary.up',
+              },
+            },
+            query,
+          ],
+        },
+      },
+      sort: [
+        {
+          '@timestamp': 'desc',
+        },
+      ],
+      size: 0,
+      aggs: {
+        monitors: {
+          composite: {
+            size: LEGACY_STATES_QUERY_SIZE,
+            sources: [
+              {
+                monitor_id: {
+                  terms: {
+                    field: 'monitor.id',
+                  },
+                },
+              },
+            ],
+          },
+          aggs: {
+            top: {
+              top_hits: {
+                sort: [
+                  {
+                    '@timestamp': 'desc',
+                  },
+                ],
+                _source: {
+                  includes: ['monitor.check_group', '@timestamp', 'agent.id'],
+                },
+                // The idea here is that we want to get enough documents to get all
+                // possible agent IDs. Doing that in a deterministic way is hard,
+                // but all agent IDs should be represented in the top 50 results in most cases.
+                // There's an edge case here where a user has accidentally configured
+                // two agents to run on different schedules, but that's an issue on the user side.
+                size: 50,
+              },
+            },
+          },
+        },
+      },
+    };
+
+    if (searchAfter) {
+      set(body, 'aggs.monitors.composite.after', searchAfter);
+    }
+
+    const params = {
+      index: INDEX_NAMES.HEARTBEAT,
+      body,
+    };
+
+    const result = await this.database.search(request, params);
+
+    const checkGroups = result.aggregations.monitors.buckets.flatMap((b: any) => {
```

Comment:
Can we rename `b` to `bucket`? I know this is subjective and that's a fairly standard pattern but one-letter variable names always seem obscure to me.